amqp/types.py

```diff
@@ -0,0 +1,566 @@
+    connected: bool = False
+    host: str
+    port: int
+    ssl: Any
+    connect_timeout: float
+    read_timeout: float
+    write_timeout: float
+    socket_settings: Mapping
+    sock: socket.socket
+
+    @abc.abstractmethod
+    async def connect(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    def close(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def read_frame(self, timeout: float = None) -> Frame:
+        ...
+
+    @abc.abstractmethod
+    async def write(self, s: bytes, timeout: float = None) -> None:
+        ...
+
+
+class ContentT(metaclass=abc.ABCMeta):
+    """Generic content type."""
+
+    CLASS_ID: int
+    PROPERTIES: Sequence[Tuple[str, str]]
+
+    properties: MutableMapping
+    body_received: int = 0
+    body_size: int = 0
+    ready: bool = False
+
+    frame_method: method_sig_t
+    frame_args: str
+
+    @abc.abstractmethod
+    def _load_properties(
+            self,
+            class_id: int,
+            buf: bytes,
+            offset: int = 0,
+            classes: Mapping = None,
+            unpack_from: Callable = None) -> int:
+        ...
+
+    @abc.abstractmethod
+    def _serialize_properties(self) -> bytes:
+        ...
+
+    @abc.abstractmethod
+    def inbound_header(self, buf: bytes, offset: int = 0) -> int:
+        ...
+
+    @abc.abstractmethod
+    def inbound_body(self, buf: bytes):
+        ...
+
+
+class MessageT(ContentT, metaclass=abc.ABCMeta):
+    """Basic message type."""
+
+    body: bytes
+    children: Any
+    channel: 'ChannelT'
+    delivery_info: Mapping[str, Any]
+
+    content_type: str
+    content_encoding: str
+    application_headers: MutableMapping
+    delivery_mode: int
+    priority: int
+    correlation_id: str
+    reply_to: str
+    expiration: str
+    message_id: str
+    timestamp: datetime
+    type: str
+    user_id: str
+    app_id: str
+    cluster_id: str
+
+    @abc.abstractmethod
+    def __init__(self,
+                 body: bytes=b'',
+                 *,
+                 children: Any = None,
+                 channel: 'ChannelT' = None) -> None:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def headers(self) -> MutableMapping:
+        ...
+
+    @property
+    @abc.abstractmethod
+    def delivery_tag(self) -> str:
+        ...
+
+
+class AbstractChannelT(metaclass=abc.ABCMeta):
+    """Abstract channel type."""
+
+    connection: 'ConnectionT'
+    channel_id: int
+    auto_decode: bool = False
+    is_open: bool = False
+
+    @abc.abstractmethod
+    def __enter__(self) -> 'AbstractChannelT':
+        ...
+
+    @abc.abstractmethod
+    def __exit__(self, *exc_info) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def __aenter__(self) -> 'AbstractChannelT':
+        ...
+
+    @abc.abstractmethod
+    async def __aexit__(self, *exc_info) -> None:
+        ...
+
+    @abc.abstractmethod
+    def _setup_listeners(self):
+        ...
+
+    @abc.abstractmethod
+    async def send_method(
```

Comment:
should we introduce async methods or start with sync only for now?

Reply:
This is a design question we need to answer before working on this PR.
As far as I'm concerned, we should drop support for synchronous connections as it will make it harder to maintain this library.

Reply:
OK. thanks